### PR TITLE
Move the Discharge module in the kernel and merge it with Cooking

### PR DESCRIPTION
**Pull request by ppedrot**
This module is undoubtedly part of the TCB, but it was living in the upper strata. We merge it with Cooking, as they share a common function (and similar code).

Ultimately to escape from soundness holes like #8082, the kernel should be responsible for the logical handling of sections, but it is a much bigger task that I'll try to implement at some later point.

### CODE REVIEW

**Review by maximedenes — APPROVED** [review]

**Comment by maximedenes** [comment]
In the unlikely event this change may appear as controversial, I'll wait a bit. In particular, I'd like other reviewers to have a look and approve or request changes.

File: kernel/cooking.ml [hunk 1]
```diff
@@ -185,32 +185,37 @@ let cook_constr { Opaqueproof.modlist ; abstract = (vars, subst, _) } c =
   let hyps = abstract_context hyps in
   abstract_constant_body (expmod c) hyps
 
+let discharge_abstract_universe_context subst abs_ctx auctx =
```
Comment (SkySkimmer) [inline kernel/cooking.ml]: The copy of this function in Lib should be deleted.

File: kernel/cooking.ml [hunk 2]
```diff
@@ -251,4 +256,115 @@ let cook_constant ~hcons { from = cb; info } =
...
+     Var(y1)..Var(yq):C1..Cq; I1..Ip:B1..Bp |- ci : Ti
+
+   by
+
+     |- Ij: (y1..yq:C1..Cq)Bj
+     I1..Ip:(B1 y1..yq)..(Bp y1..yq) |- ci : (y1..yq:C1..Cq)Ti[Ij:=(Ij y1..yq)]
+*)
+
+let it_mkNamedProd_wo_LetIn b d =
+  List.fold_left (fun c d -> mkNamedProd_wo_LetIn d c) b d
```
Comment (SkySkimmer) [inline kernel/cooking.ml]: How about moving this to Term and removing the copy in Termops?

File: kernel/cooking.ml [hunk 3]
```diff
@@ -251,4 +256,115 @@ let cook_constant ~hcons { from = cb; info } =
...
+    | Polymorphic auctx ->
+      let subst, auctx = discharge_abstract_universe_context subst abs_uctx auctx in
+      let subst = Univ.make_instance_subst subst in
+      let nas = Univ.AUContext.names auctx in
+      let auctx = Univ.AUContext.repr auctx in
+      subst, Polymorphic_entry (nas, auctx)
+  in
+  let variance = match mib.mind_variance with
+    | None -> None
+    | Some _ -> Some (dummy_variance ind_univs)
```
Comment (SkySkimmer) [inline kernel/cooking.ml]: I don't know if dummy_variance is a great idea. It looks like it should produce a valid variance for the inductive, but if that was true it should use Invariant not Irrelevant. Maybe we should just leave the original variance and let rebuild deal with it (note that infercumul only tests whether it's a Some without looking at the content).

**Review by SkySkimmer — APPROVED** [review]

File: kernel/cooking.ml (hunk 3 above)
Comment (ppedrot) [reply kernel/cooking.ml]: I am not changing this in this PR, you should probably open a new PR atop of this simple merge.

File: kernel/cooking.ml (hunk 2 above)
Comment (ppedrot) [reply kernel/cooking.ml]: That would make the API asymmetrical, there are other similar functions in Termops not used in the kernel.

File: kernel/cooking.ml (hunk 1 above)
Comment (ppedrot) [reply kernel/cooking.ml]: It's used by Classes though?

Comment (SkySkimmer) [reply kernel/cooking.ml]: So export it from cooking

Comment (ppedrot) [reply kernel/cooking.ml]: Classes seems to reimplement its own cooking part. I am not sure I want to export internal stuff that is simple enough to be done locally...

Comment (SkySkimmer) [reply kernel/cooking.ml]: discharge_abstract_universe_context is not simple
If it was really internal there wouldn't be a copy in lib
Don't leave copies around after moving code

**Review by SkySkimmer — CHANGES_REQUESTED** [review]
Please no duplicates of discharge_abstract_universe_context, I never fully understood this function so I don't want to have to fail to understand it twice

**Comment by ppedrot** [comment]
I don't agree with the exposure of `discharge_abstract_universe_context` in Cooking. The signature of the kernel-side function is unreadable, while the one in `Lib` uses the private `abstr_info` type, ensuring that it is not misused. Thus the Lib one provides more static invariants checked by OCaml.

**Comment by maximedenes** [comment]
@ppedrot IIUC what you'd like to do is to have `Lib` use the `discharge_abstract_universe_context` from `Cooking`, but giving it a more abstract signature and sealing the whole thing from the outside. AFAIK, OCaml doesn't really support that, the idiomatic way would be to export the function from `Cooking` and document that external users should look at `Lib` instead. What's the problem with that?

Duplicating code is certainly not the way to go.

**Comment by ppedrot** [comment]
> What's the problem with that?

Kernel API bloat.

> Duplicating code is certainly not the way to go.

I am *not* duplicating the code, it's already written this way.

**Comment by maximedenes** [comment]
> Kernel API bloat.

I think the API looks ok, even with this function exported.

> I am _not_ duplicating the code, it's already written this way.

Ok, but I don't think it should be the case. Whether you want to address it now or not, I don't really care.

**Review by SkySkimmer — APPROVED** [review]
If you have a strong opinion to keep it then let's move on

File: kernel/cooking.ml [hunk 4]
```diff
@@ -185,32 +185,37 @@ let cook_constr { Opaqueproof.modlist ; abstract = (vars, subst, _) } c =
...
-    *)
-    if (Univ.Instance.is_empty subst) then
-      (** Still need to take the union for the constraints between globals *)
-      subst, (Polymorphic (AUContext.union auctx0 auctx))
-    else
-      let ainst = Univ.make_abstract_instance auctx in
-      let subst = Instance.append subst ainst in
-      let substf = Univ.make_instance_subst subst in
-      let auctx' = Univ.subst_univs_level_abstract_universe_context substf auctx in
-      subst, (Polymorphic (AUContext.union auctx0 auctx'))
```
Comment (SkySkimmer) [inline kernel/cooking.ml]: I guess you see it as moving this block rather than duplicating the lib function

**Comment by maximedenes** [comment]
@herbelin do you approve this PR?

**Review by herbelin — COMMENTED** [review]
I'm strongly approving the direction taken by this PR. I'm resting on the discussion between the author and proper reviewers for the details.

**Pull request merged**